odl/tomo/operators/ray_trafo.py

```diff
@@ -285,9 +186,96 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
                                                    proj_space.dtype,
                                                    reco_space.dtype))
 
+        if reco_space.ndim != geometry.ndim:
+            raise ValueError('`{}.ndim` not equal to `geometry.ndim`: '
+                             '{} != {}'.format(reco_name, reco_space.ndim,
+                                               geometry.ndim))
+
+        # Cache for input/output arrays of transforms
+        self.use_cache = kwargs.pop('use_cache', True)
+
+        impl = kwargs.pop('impl', None)
+        self.__cached_impl = None
+
+        if impl is None:  # User didn't specify a backend
+            if not _AVAILABLE_IMPLS:
+                raise RuntimeError('no ray transform back-end available; '
+                                   'this requires 3rd party packages, please '
+                                   'check the install docs')
+
+            # Select fastest available, _AVAILABLE_IMPLS is sorted by speed
+            impl_type = _SUPPORTED_IMPL[_AVAILABLE_IMPLS[0]]
+
+            # Warn if implementation is not fast enough
+            if not impl_type.can_handle_size(reco_space.size):
+                if impl_type == AstraCpuRayTransformImpl:
+                    warnings.warn(
+                        "The best available backend ('astra_cpu') may be too "
+                        "slow for volumes of this size. Consider using "
+                        "'astra_cuda' if your machine has an Nvidia GPU. "
+                        "This warning can be disabled by explicitly setting "
+                        "`impl='astra_cpu'`.",
+                        RuntimeWarning)
+                elif impl_type == SkimageRayTransformImpl:
+                    warnings.warn(
+                        "The best available backend ('skimage') may be too "
+                        "slow for volumes of this size. Consider using ASTRA. "
+                        "This warning can be disabled by explicitly setting "
+                        "`impl='skimage'`.",
+                        RuntimeWarning)
```

Comment:
Why this abstraction leak? In principle, the warning text could be part of the implementation class, so `RayTransformBase` wouldn't have to know about specific impls.

Reply:
The motivation is to keep the same error messages, and a backend class is unaware of the others. The abstraction doesn't leak I believe, `RayTransform` is just aware of special cases. I changed the error messages a bit and moved this into the `__init__` of the backends as well.